go/libkb/pvl_interp.go

```diff
@@ -0,0 +1,825 @@
+
+	selectorIndex, err := selector.GetInt()
+	selectorIsIndex := err == nil
+	selectorKey, err := selector.GetString()
+	selectorIsKey := err == nil && !selectorIsIndex
+	allness, err := selector.AtKey("all").GetBool()
+	selectorIsAll := err == nil && allness
+
+	switch {
+	case selectorIsIndex:
+		object, err := object.ToArray()
+		if err != nil {
+			g.GetLog().Debug("PVL json select by index from non-array: %v", err)
+			return []string{}, nil
+		}
+
+		nextobject := object.AtIndex(selectorIndex)
+		return pvlRunSelectorJSONInner(g, nextobject, nextselectors)
+	case selectorIsKey:
+		object, err := object.ToDictionary()
+		if err != nil {
+			g.GetLog().Debug("PVL json select by key from non-map: %v", err)
+			return []string{}, nil
+		}
+
+		nextobject := object.AtKey(selectorKey)
+		return pvlRunSelectorJSONInner(g, nextobject, nextselectors)
+	case selectorIsAll:
+		children, err := pvlJSONGetChildren(object)
+		if err != nil {
+			g.GetLog().Debug("PVL json select could not get children: %v", err)
+			return []string{}, nil
+		}
+		var results []string
+		for _, child := range children {
+			innerresults, perr := pvlRunSelectorJSONInner(g, child, nextselectors)
+			if perr != nil {
+				return nil, perr
+			}
+			results = append(results, innerresults...)
+		}
+	}
+	return []string{}, NewProofError(keybase1.ProofStatus_INVALID_PVL,
+		"Selector entry not recognized: %v", selector)
+}
+
+// Take a template, substitute variables, and build the Regexp.
+func pvlInterpretRegex(template string, vars PvlScriptVariables) (*regexp.Regexp, ProofError) {
+	perr := NewProofError(keybase1.ProofStatus_INVALID_PVL,
+		"Could not build regex %v", template)
+
+	// Parse out side bars and option letters.
+	if !strings.HasPrefix(template, "/") {
+		return nil, perr
+	}
+	lastSlash := strings.LastIndex(template, "/")
+	if lastSlash == -1 {
+		return nil, perr
+	}
+	opts := template[lastSlash+1:]
+	if !regexp.MustCompile("[imsU]*").MatchString(opts) {
+		return nil, perr
+	}
+	var prefix = ""
+	if len(opts) > 0 {
+		prefix = "(?" + opts + ")"
+	}
+
+	// Do variable interpolation.
+	prepattern, err := pvlSubstitute(template[1:lastSlash], vars, nil)
+	if err != nil {
+		return nil, NewProofError(keybase1.ProofStatus_BAD_API_URL, err.Error())
+	}
+	pattern := prefix + prepattern
+
+	// Build the regex.
+	re, err := regexp.Compile(pattern)
+	if err != nil {
+		return nil, NewProofError(keybase1.ProofStatus_INVALID_PVL,
+			"Could not compile regex (%v): %v", template, err)
+	}
+	return re, nil
+}
+
+// Substitute vars for %{name} in the string.
+// Only substitutes whitelisted variables.
+// It is an error to refer to an undefined numbered group. But not an error to refer to a poorly-named variable.
+// Match is an optional slice which is a regex match.
+func pvlSubstitute(template string, vars PvlScriptVariables, match []string) (string, error) {
+	var outerr error
+	// Regex to find %{name} occurrences.
+	re := regexp.MustCompile("%{[\\w0-9]+}")
```

Comment:
Looks like `\w` already includes the digits? https://golang.org/pkg/regexp/syntax/